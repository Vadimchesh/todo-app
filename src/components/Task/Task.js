import React from 'react';
import styles from './style.module.scss';
import cn from 'classnames';
import { useDispatch } from 'react-redux';

const TodoListItem = ({ id, text, completed }) => {
  const dispatch = useDispatch();

  const handleCompletedChanged = () => {
    dispatch({ type: 'todos/todoToggle', payload: id });
  };

  const onDelete = () => {
    dispatch({ type: 'todos/todoDelet', payload: id });
  };

  return (
    <label data-cy='labelTaskText' className={styles.label}>
      <div className={styles.container}>
        <input
          data-cy='toogleTask'
          className='toggle'
          type='checkbox'
          checked={completed}
          onChange={handleCompletedChanged}
        />
        <span className={styles.checkmark}></span>
        <span
          data-cy='taskText'
          className={cn({
            [styles.done]: completed,
          })}
        >
          {text}
        </span>
        <span
          data-cy='deleteBtn'
          id='delete'
          className={styles.close}
          onClick={onDelete}
        ></span>
      </div>
    </label>
  );
};
export default TodoListItem;
