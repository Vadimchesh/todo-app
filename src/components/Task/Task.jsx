import React from "react";
import styles from "./style.module.scss";
import cn from "classnames";
import { useSelector, useDispatch } from "react-redux";

const selectTodoById = (state, todoId) => {
  return state.todos.find((todo) => todo.id === todoId);
};

const TodoListItem = ({ id }) => {
  const todo = useSelector((state) => selectTodoById(state, id));
  const { text, completed } = todo;

  const dispatch = useDispatch();

  const handleCompletedChanged = () => {
    dispatch({ type: "todos/todoToggled", payload: todo.id });
  };

  const onDelete = () => {
    dispatch({ type: "todos/todoDeleted", payload: todo.id });
  };

  return (
    <label className={styles.label}>
      <div className={styles.container}>
        <input
          className="toggle"
          type="checkbox"
          checked={completed}
          onChange={handleCompletedChanged}
        />
        <span className={styles.checkmark}></span>
        <span
          className={cn({
            [styles.done]: todo.completed,
          })}
        >
          {text}
        </span>
      </div>
      <button className={styles.del} onClick={onDelete}></button>
    </label>
  );
};

export default TodoListItem;
