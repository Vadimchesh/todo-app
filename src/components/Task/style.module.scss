@import '../../assets/scss/variables.scss';

.container {
  position: relative;
  @include container;
  &:hover input ~ .checkmark {
    background-color: rgb(243, 241, 241);
  }
  input {
    position: absolute;
    margin-right: 1rem;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;
    &:checked + .checkmark:after {
      opacity: 1;
      border-color: $done-color;
    }
    &:checked ~ .checkmark {
      border-color: $done-color;
    }
  }
}
.checkmark {
  position: relative;
  margin-right: 1rem;
  background-color: rgb(252, 252, 252);
  border: 1px solid $gray-color;
  border-radius: 50%;
  height: 1.75rem;
  width: 1.75rem;
  &:after {
    content: '';
    position: absolute;
    top: 0.5rem;
    left: 0.45rem;
    opacity: 0;
    border: 2px solid $done-color;
    border-top: none;
    border-right: none;
    transform: rotate(-45deg);
    width: 0.75rem;
    height: 0.375rem;
  }
}

.done {
  text-decoration: line-through;
  color: $gray-color;
}

.close {
  position: absolute;
  width: 32px;
  height: 32px;
  left: 90%;
  opacity: 0.3;
}
.close:hover {
  opacity: 1;
}
.close:before,
.close:after {
  position: absolute;
  left: 15px;
  content: ' ';
  height: 33px;
  width: 2px;
  background-color: $red-color;
}
.close:before {
  transform: rotate(45deg);
}
.close:after {
  transform: rotate(-45deg);
}

.label {
  position: relative;
}
