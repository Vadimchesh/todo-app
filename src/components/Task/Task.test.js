import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { useDispatch } from 'react-redux';
import Task from './Task';

const mockDispatch = jest.fn();

jest.mock('react-redux', () => {
  const originalModule = jest.requireActual('react-redux');
  return {
    ...originalModule,
    useDispatch: jest.fn(() => mockDispatch),
  };
});

describe('Task:', () => {
  let wrapper = ShallowWrapper;
  beforeEach(() => {
    wrapper = shallow(<Task />);
  });
  afterAll(() => {
    jest.resetAllMocks();
  });

  it('snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('called toggle fn', () => {
    wrapper.find('.toggle').simulate('change');
    expect(mockDispatch).toBeCalled();
  });

  it('called onDelete fn', () => {
    wrapper.find('#delete').simulate('click');
    expect(mockDispatch).toBeCalled();
  });
});
