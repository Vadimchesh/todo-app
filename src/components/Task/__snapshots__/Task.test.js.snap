// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task: snapshot 1`] = `
<label
  className="label"
  data-cy="labelTaskText"
>
  <div
    className="container"
  >
    <input
      className="toggle"
      data-cy="toogleTask"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="checkmark"
    />
    <span
      className=""
      data-cy="taskText"
    />
    <span
      className="close"
      data-cy="deleteBtn"
      id="delete"
      onClick={[Function]}
    />
  </div>
</label>
`;
